app/src/main/kotlin/com/wire/android/migration/MigrationManager.kt

```diff
@@ -117,28 +115,36 @@ class MigrationManager @Inject constructor(
         updateProgress: suspend (MigrationData.Progress) -> Unit,
     ): MigrationData.Result =
         try {
+            val report: MigrationReport = MigrationReport()
             updateProgress(MigrationData.Progress(MigrationData.Progress.Type.MESSAGES))
-            appLogger.d("$TAG - Step 3 - Migrating users for ${userId.value.obfuscateId()}")
-            migrateUsers(userId).flatMap {
-                appLogger.d("$TAG - Step 4 - Migrating conversations for ${userId.value.obfuscateId()}")
-                migrateConversations(it)
-            }.flatMap {
-                appLogger.d("$TAG - Step 5 - Migrating messages for ${userId.value.obfuscateId()}")
-                migrateMessages(userId, it, coroutineScope).let { failedConversation ->
-                    if (failedConversation.isEmpty()) {
-                        Either.Right(Unit)
-                    } else {
-                        Either.Left(failedConversation.values.first())
-                    }
+            appLogger.d("$TAG - Step 1 - Migrating users for ${userId.value.obfuscateId()}")
+            migrateUsers(userId)
+                .also { report.addUserReport(userId, it) }
+                .flatMap {
+                    appLogger.d("$TAG - Step 2 - Migrating conversations for ${userId.value.obfuscateId()}")
+                    migrateConversations(it).also { report.addConversationReport(userId, it) }
                 }
-            }.fold({
-                when (it) {
-                    is NetworkFailure.NoNetworkConnection -> MigrationData.Result.Failure.NoNetwork
-                    else -> MigrationData.Result.Failure.Messages(it.getErrorCode().toString())
+                .flatMap {
+                    appLogger.d("$TAG - Step 3 - Migrating messages for ${userId.value.obfuscateId()}")
```

Comment:
Can we some how extract all those functions to make it reusable in `migrateSingleUser` `migrateAccountsData`